Add documentation about deprecating Leap features (#377)
dwavesystems/dwave-ocean-sdk: docs/leap_sapi/index.rst

Up to 1000 of your most recent problems are stored and accessible in the Leap
service for up to 365 days; if the number of your problems exceeds 1000, the
Leap service begins to delete the oldest ones.

.. _deprecated_features:

> Feature Deprecation

Reviewer: Refer to the following for a list of fields and description: 
https://dwavesys.atlassian.net/wiki/spaces/SSW/pages/1437696081/Technical+design+for+deprecation+message+propagation+from+Leap+to+Ocean+-+FINAL

> Feature Deprecation

Reviewer 2: @qwriter please note that the dates (@1717171200 and @1717171200) are still made up and that this is not deployed to PROD

> Feature Deprecation

Reviewer: @sjain-dwave : are you planning to fix the Accept header case issue prior to release next week? https://git.dwavesys.local/projects/AUT/repos/terraform/pull-requests/2519/overview?commentId=75743

===================

A Leap service feature may be deprecated and, during the deprecation period,
information about such a deprecation is displayed whenever the feature is used.
For example, a message would be displayed when a deprecated SAPI endpoint is
called. For the Ocean SDK, deprecation messages are displayed in a terminal; for
SAPI, deprecation messages are displayed as HTTP ``X-Deprecation`` response
headers. A deprecation message includes the following:

> *   Brief description

Reviewer 3: I think that similar to your description of ``context``, the format of this part of the message needs to be described too, no?

> *   Deprecation ID, which can be used to find a more detailed description

Reviewer 3: I suggest just "can be used to find the relevant release note"

for some deprecations the notice will contain sufficient information and the RN will not add to it

> *   Deprecation ID, which can be used to find a more detailed description

Reviewer 3: I think that similar to your description of ``context``, the format of this part of the message needs to be described too, no?

    in the release notes

*   ``context`` string, which identifies the affected Leap service category as
    follows:

>     *   ``api``: SAPI, such as endpoints, HTTP headers, and data structures

Reviewer 3: It may be the bullet formatting but it's unclear to me whether the ``context`` string is like a dict that contains all these 5 "fields" with each having a "value" or if it looks like "context: parameter".

Reviewer 2: Yes, the value is just a string. It looks like this:
context="api"
I think it will be a lot clearer if you provide an example like @randomir suggested in this comment:
[https://github.com/dwavesystems/dwave-ocean-sdk/pull/377#discussion_r2322574805](https://github.com/dwavesystems/dwave-ocean-sdk/pull/377#discussion_r2322574805)

    *   ``feature``: Solver features

    *   ``parameter``: Solver parameters

    *   ``solver``: QPU solvers

>     *   ``solver type``: Hybrid solvers

Reviewer 3: I don't understand why ``solver`` would mean QPU and ``solver type`` a hybrid solver. Is this correct?

> *   Link to the release notes, which includes a more detailed description about
>     the deprecation

Reviewer 3: ```suggestion
*   Link to the release note for the deprecation
```

> *   Sunset date, which is the date after which the feature is no longer
>     guaranteed to be supported

Reviewer 3: ```suggestion
*   Sunset date, after which date support for the feature is no longer
    guaranteed
```
